src/raft.rs

```diff
@@ -2215,14 +2450,93 @@ impl<T: Storage> Raft<T> {
         self.lead_transferee = None;
     }
 
+    // Send snapshot request message to the leader
     fn send_request_snapshot(&mut self) {
+        let m = self.prepare_request_snapshot(self.leader_id);
+        self.send(m);
+    }
+
+    // Prepare a Message for requesting snapshot from given `to`
+    fn prepare_request_snapshot(&mut self, to: u64) -> Message {
         let mut m = Message::default();
         m.set_msg_type(MessageType::MsgAppendResponse);
         m.index = self.raft_log.committed;
         m.reject = true;
         m.reject_hint = self.raft_log.last_index();
-        m.to = self.leader_id;
+        m.to = to;
         m.request_snapshot = self.pending_request_snapshot;
-        self.send(m);
+        m
+    }
+
+    #[inline]
+    fn pr_become_snapshot_and_log(
+        &self,
+        id: u64,
+        progress: &mut Progress,
+        snapshot_index: u64,
+        snapshot_term: u64,
+    ) {
+        debug!(
+            self.logger,
+            "[firstindex: {first_index}, commit: {committed}] sent snapshot[index: {snapshot_index}, term: {snapshot_term}] to {to}",
+            first_index = self.raft_log.first_index(),
+            committed = self.raft_log.committed,
+            snapshot_index = snapshot_index,
+            snapshot_term = snapshot_term,
+            to = id;
+            "progress" => ?progress,
+        );
+        progress.become_snapshot(snapshot_index);
+        debug!(
+            self.logger,
+            "paused sending replication messages to {}",
+            id;
+            "progress" => ?progress,
+        );
+    }
+
+    // Pick a delegate of the given group
+    //
+    // The delegate must satisfy conditions below:
+    // 1. Must be 'recent_active'
+    // 2. The progress state should be 'Replicate' but not 'paused'
```

Comment:
Seems `Probe` needs also to be allowed?

Reply:
`Probe` might not be suitable for a delegate in the leader's view because the `Probe` member can probably just recover from a crash or is suffering the network partition.